shared/x509utils/types.go

```diff
@@ -27,3 +30,40 @@ type PrivateKey interface {
 type PublicKey interface {
 	Equal(x crypto.PublicKey) bool
 }
+
+// BlockToPrivateKey parses a pem Block looking for rsa, ecdsa or ed25519 Private Keys
+func BlockToPrivateKey(block *pem.Block) (PrivateKey, error) {
+	if block.Type == "PRIVATE KEY" || strings.HasSuffix(block.Type, " PRIVATE KEY") {
+		if pk, _ := x509.ParsePKCS1PrivateKey(block.Bytes); pk != nil {
+			return pk, nil
+		} else if pk, err := x509.ParsePKCS8PrivateKey(block.Bytes); err != nil {
+			return nil, err
+		} else if key, ok := pk.(PrivateKey); ok {
+			return key, nil
+		}
+	}
+	return nil, nil
+}
+
+// BlockToRSAPrivateKey attempts to parse a pem.Block to extract an rsa.PrivateKey
+func BlockToRSAPrivateKey(block *pem.Block) (*rsa.PrivateKey, error) {
+	if pk, err := BlockToPrivateKey(block); err != nil {
+		return nil, err
+	} else if key, ok := pk.(*rsa.PrivateKey); ok {
+		return key, nil
+	} else {
+		return nil, nil
+	}
+}
+
+// BlockToCertificate attempts to parse a pem.Block to extract a x509.Certificate
+func BlockToCertificate(block *pem.Block) (*x509.Certificate, error) {
+	if block.Type == "CERTIFICATE" {
+		if cert, err := x509.ParseCertificate(block.Bytes); err != nil {
+			return nil, err
+		} else if cert != nil {
+			return cert, nil
+		}
+	}
+	return nil, nil
```

Comment:
same as previous, nil, nil is confusing, for me it means "there was no error but also no result" mistery.....

Reply:
see other comments on the same topic

Reply:
Yup